### PR TITLE
🌱Add test for baremetalhost controller updateEventHandler

### PR DESCRIPTION
**Pull request by troy0820**
**What this PR does / why we need it**:

~~When using this method, the `saveHostStatus` will requeue the object and create more reconciles for the object when the only thing that has changed is the lastupdated field in the status.  This leads to multiple reconciles because this happens during each run through the reconciler.~~

This adds a test to the baremetalhost controller to validate that the saveHostStatus will not requeue the object and create more reconciles when lastUpdated is updated.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
~~Fixes https://github.com/metal3-io/baremetal-operator/issues/1253~~

### CODE REVIEW

**Comment by troy0820** [comment]
Actually it looks like it may be in here:
```go
_, oldOK := e.ObjectOld.(*metal3api.BareMetalHost)
	_, newOK := e.ObjectNew.(*metal3api.BareMetalHost)
	if !(oldOK && newOK) {
		// The thing that changed wasn't a host, so we
		// need to assume that we must update. This
		// happens when, for example, an owned Secret
		// changes.
		return true
	}

```
Because the status will not be the same and then it will return true and not process the rest of the handler.

File: controllers/metal3.io/baremetalhost_controller.go [hunk 1]
```diff
@@ -2080,7 +2080,7 @@ func (r *BareMetalHostReconciler) updateEventHandler(e event.UpdateEvent) bool {
 	}
 
 	// Discard updates that did not increase the resource Generation (such as on Status.LastUpdated), except for the finalizers or annotations
-	if reflect.DeepEqual(e.ObjectNew.GetFinalizers(), e.ObjectOld.GetFinalizers()) && reflect.DeepEqual(e.ObjectNew.GetAnnotations(), e.ObjectOld.GetAnnotations()) {
+	if reflect.DeepEqual(e.ObjectNew.GetFinalizers(), e.ObjectOld.GetFinalizers()) && reflect.DeepEqual(e.ObjectNew.GetAnnotations(), e.ObjectOld.GetAnnotations()) || oldBMH.Status.LastUpdated != newBMH.Status.LastUpdated {
 		return false
```
Comment (troy0820) [inline controllers/metal3.io/baremetalhost_controller.go:2084]: This seems wrong, in the test below, we would want that event to come through because the finalizer is different from old/new, this prevents it because of the or. We can move this to a new line and isolate that.

Comment (troy0820) [reply controllers/metal3.io/baremetalhost_controller.go]: Added test and removed the logic on the line with the false condition.  Now this will be isolated from the event where the LastUpdated was modified but nothing else changed.

**Comment by tuminoid** [comment]
Thanks for the contribution! Let's squash those commits up.

Let's get some eyes on this.
/cc @kashifest @Rozzii @dtantsur @elfosardo

**Comment by kashifest** [comment]
I am trying to understand this patch a bit, the original issue pointed by @s3rj1k was an edge case where code updated the status/BMH object even when there was no actual change in status, but it seems this patch is ignoring updating the object even when theres is any update on the status? Am I understanding it correctly?

File: controllers/metal3.io/baremetalhost_controller.go [hunk 2]
```diff
@@ -2079,6 +2079,11 @@ func (r *BareMetalHostReconciler) updateEventHandler(e event.UpdateEvent) bool {
 		return true
 	}
 
+	// The assumption that this will not always be filled out is why this check is not appended to the below block
+	if oldBMH.Status.LastUpdated != nil && newBMH.Status.LastUpdated != nil && oldBMH.Status.LastUpdated != newBMH.Status.LastUpdated && reflect.DeepEqual(oldBMH.Spec, newBMH.Spec) && reflect.DeepEqual(oldBMH.ObjectMeta, newBMH.ObjectMeta) {
```
Comment (zaneb) [inline controllers/metal3.io/baremetalhost_controller.go:2083]: Why only skip reconcile when LastUpdated has changed?

Naively I'd have thought we'd want to skip it _even more_ when LastUpdated has **not** changed.

File: controllers/metal3.io/baremetalhost_controller.go [hunk 3]
```diff
@@ -2079,6 +2079,11 @@ func (r *BareMetalHostReconciler) updateEventHandler(e event.UpdateEvent) bool {
...
+		return false
+	}
+
 	// Discard updates that did not increase the resource Generation (such as on Status.LastUpdated), except for the finalizers or annotations
 	if reflect.DeepEqual(e.ObjectNew.GetFinalizers(), e.ObjectOld.GetFinalizers()) && reflect.DeepEqual(e.ObjectNew.GetAnnotations(), e.ObjectOld.GetAnnotations()) {
```
Comment (zaneb) [inline controllers/metal3.io/baremetalhost_controller.go:2088]: Isn't this already a rather awkward way of saying `if reflect.DeepEqual(oldBMH.ObjectMeta, newBMH.ObjectMeta) { return false }`?
i.e. this already captures everything that the new test above would capture, and then some.

**Comment by troy0820** [comment]
> I am trying to understand this patch a bit, the original issue pointed by @s3rj1k was an edge case where code updated the status/BMH object even when there was no actual change in status, but it seems this patch is ignoring updating the object even when theres is any update on the status? Am I understanding it correctly?

The object is updated but it doesn’t requeue the object in the event that the status last changed gets updated during a saveStatus. This was supposed to be handled by the predicate generation changed but seems to slip through.

**Review by zaneb — COMMENTED** [review]
I'm not sure this has any effect.

It's probably safe as long as we always remember to return Requeue:true whenever we need to requeue.
I think we do, but I also thought that we avoid writing unchanged statuses (as Dmitry [pointed out](https://github.com/metal3-io/baremetal-operator/issues/1253#issuecomment-1549695960)), yet according the the bug report we do not. (OTOH, I'm not sure whether to believe this, because when asked where this happens the [response](https://github.com/metal3-io/baremetal-operator/issues/1253#issuecomment-1647654820) was "check the code", and yet upon checking the code it's [clearly fine](https://github.com/metal3-io/baremetal-operator/blob/main/controllers/metal3.io/baremetalhost_controller.go#L233-L237).)

File: controllers/metal3.io/baremetalhost_controller.go (hunk 3 above)
Comment (troy0820) [reply controllers/metal3.io/baremetalhost_controller.go:2088]: The above check from this patch would make sure that everything in ObjectMeta is equal including what is not explicitly set with the Finalizers and/or the Annotations, e.g: Labels, OwnerRefs, ResourceVersion, Generation, etc.  

The check on line 2088 seems to only care about the Finalizers and Annotations which would create requeues with Label changes, ownerRefs, etc which would be valid. 

The check I am putting in makes sure that the only change that is different is that the LastUpdated on the status is changed without anything else being different, therefor not requeuing.

**Comment by troy0820** [comment]
Running the test that I added as part of this change works without the patch included.  So I can dismiss this change and leave the test but we probably should close the issue if it is fixed.

File: controllers/metal3.io/baremetalhost_controller_test.go [hunk 4]
```diff
@@ -1766,6 +1769,54 @@ func TestUpdateEventHandler(t *testing.T) {
 
 			expectedProcess: true,
 		},
+		{
+			name: "do-requeue-when-finalizers-are-different",
```
Comment (andfasano) [inline controllers/metal3.io/baremetalhost_controller_test.go:1773]: Isn't already covered by https://github.com/metal3-io/baremetal-operator/pull/1881/files#diff-20c1a976622b8d54cf98349e3e6219778fa00e89ed669f1bb12cd778326469bbR1733?

File: controllers/metal3.io/baremetalhost_controller_test.go [hunk 5]
```diff
@@ -1766,6 +1769,54 @@ func TestUpdateEventHandler(t *testing.T) {
...
+				},
+					Status: metal3api.BareMetalHostStatus{
+						LastUpdated: &metav1.Time{Time: later},
+					}},
+			},
+
+			expectedProcess: true,
+		},
+		{
+			name: "do-not-requeue-when-status-lastupdated-are-different",
```
Comment (andfasano) [inline controllers/metal3.io/baremetalhost_controller_test.go:1797]: Just my two cents, but given that there is already a test for [base case](https://github.com/metal3-io/baremetal-operator/pull/1881/files#diff-20c1a976622b8d54cf98349e3e6219778fa00e89ed669f1bb12cd778326469bbR1693), I would find a better name something like `do-not-requeue-when-only-status-lastupdated-are-different`

Comment (troy0820) [reply controllers/metal3.io/baremetalhost_controller_test.go:1797]: I can update the name

File: controllers/metal3.io/baremetalhost_controller_test.go (hunk 4 above)
Comment (troy0820) [reply controllers/metal3.io/baremetalhost_controller_test.go:1773]: I can remove this but this was to simulate that this would requeue despite the LastUpdated in the status.  This wouldn't do anything other than prove that this still requeues and not get swallowed by the false requeue from the LastUpdated in the status.

File: controllers/metal3.io/baremetalhost_controller.go (hunk 3 above)
Comment (zaneb) [reply controllers/metal3.io/baremetalhost_controller.go]: We basically have:

```
if foo && bar && baz &&quux {
  return false
}
if foo {
  return false
}
return true
```

When put like that, it's pretty easy to see that the first check isn't doing anything.

Comment (troy0820) [reply controllers/metal3.io/baremetalhost_controller.go]: This change is now outdated and only includes the test to verify that the change implemented to address the root issue is resolved. 

The thought behind it was the first check would include the other changes that don’t get captured in the second check which is why the status is not included but only the spec/metadata and the LastUpdated.

Comment (troy0820) [reply controllers/metal3.io/baremetalhost_controller.go]: Resolving as not relevant to the review.

File: controllers/metal3.io/baremetalhost_controller.go (hunk 2 above)
Comment (troy0820) [reply controllers/metal3.io/baremetalhost_controller.go]: > Running the test that I added as part of this change works without the patch included. So I can dismiss this change and leave the test but we probably should close the issue if it is fixed.

This removes the included patch but leaves the test.  Resolving

**Comment by zaneb** [comment]
Reading over the bug report, the complaint was that we were doing no-op updates in the first place (although there is no evidence of this yet), thus causing extra work for _other_ controllers that were watching the BMH.
There was no suggestion that BMO itself was doing too many reconciles after such an update. Extra reconciles for that **and** a much broader superset of cases were already eliminated by #747.

I'm fine with adding a test though. We could even test that _all_ updates to the Status are ignored.

File: controllers/metal3.io/baremetalhost_controller_test.go (hunk 4 above)
Comment (troy0820) [reply controllers/metal3.io/baremetalhost_controller_test.go]: Resolving is the first test is removed from patch set.

File: controllers/metal3.io/baremetalhost_controller_test.go (hunk 5 above)
Comment (troy0820) [reply controllers/metal3.io/baremetalhost_controller_test.go]: The name is updated.  I am resolving.

**Review by Rozzii — COMMENTED** [review]
/lgtm

**Comment by tuminoid** [comment]
/cc @kashifest @zaneb @dtantsur @elfosardo 
Can we get an approver for this?
